PurchasesCoreSwift/Purchasing/ProductInfo.swift

```diff
@@ -0,0 +1,156 @@
+        switch paymentMode {
+        case .payUpFront:
+            return .payUpFront
+        case .payAsYouGo:
+            return .payAsYouGo
+        case .freeTrial:
+            return .freeTrial
+        default:
+            return .none
+        }
+    }
+
+    @objc public init(productIdentifier: String,
+                      paymentMode: PaymentMode,
+                      currencyCode: String,
+                      price: NSDecimalNumber,
+                      normalDuration: String?,
+                      introDuration: String?,
+                      introDurationType: RCIntroDurationType,
+                      introPrice: NSDecimalNumber?,
+                      subscriptionGroup: String?,
+                      discounts: [PromotionalOffer]?) {
+        self.productIdentifier = productIdentifier
+        self.paymentMode = paymentMode
+        self.currencyCode = currencyCode
+        self.price = price
+        self.normalDuration = normalDuration
+        self.introDuration = introDuration
+        self.introDurationType = introDurationType
+        self.introPrice = introPrice
+        self.subscriptionGroup = subscriptionGroup
+        self.discounts = discounts
+    }
+
+    @objc public func asDictionary() -> [String: NSObject] {
+        var dict: [String: NSObject] = [:]
+        dict["product_id"] = productIdentifier as NSString
+        dict["price"] = price
+        dict["currency"] = currencyCode as NSObject
+
+        if paymentMode != .none {
+            dict["payment_mode"] = NSNumber(value: paymentMode.rawValue)
+        }
+
+        if let introPrice = introPrice {
+            dict["introductory_price"] = introPrice
+        }
+
+        if let subscriptionGroup = subscriptionGroup as NSString? {
+            dict["subscription_group_id"] = subscriptionGroup
+        }
+
+        if discounts != nil {
+            dict.merge(self.discountsAsDictionary()) { (_, new) in new }
+        }
+
+        dict.merge(self.productDurationsAsDictionary()) { (_, new) in new }
+
+        return dict
+    }
+
+    @objc public var cacheKey: String {
+        var key = """
+        \(productIdentifier)-\(price)-\(currencyCode)-\(paymentMode.rawValue)-\(introPrice ?? 0)-\
+        \(subscriptionGroup ?? "")-\(normalDuration ?? "")-\(introDuration ?? "")-\(introDurationType.rawValue)
+        """
+
+        guard let discounts = discounts else {
+            return key
+        }
+
+        if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
+            for offer in discounts {
+                key += "-\(offer.offerIdentifier ?? "null offer id")"
+            }
+        }
+        return key
+    }
+
+    private func discountsAsDictionary() -> [String: NSObject] {
+        var discountDict: [String: NSObject] = [:]
+            if #available(iOS 12.2, macOS 10.14.4, tvOS 12.2, *) {
+                if let discounts = self.discounts {
+                    let offers = NSMutableArray()
```

Comment:
would using swift arrays here work?

Reply:
Same problem with using `filter/map` The type system doesn't like these dictionaries.